neuro_san/service/mcp/processors/mcp_tools_processor.py

```diff
@@ -91,11 +97,34 @@ async def call_tool(self, request_id, metadata: Dict[str, Any], tool_name: str,
         input_request: Dict[str, Any] = self._get_chat_input_request(prompt)
         response_text: str = ""
         try:
-            result_generator = service.streaming_chat(input_request, metadata)
-            async for result_dict in result_generator:
-                partial_response: str = await self._extract_tool_response_part(result_dict)
-                if partial_response is not None:
-                    response_text = response_text + partial_response
+            async with asyncio.timeout(self.tool_timeout_seconds):
+                result_generator = service.streaming_chat(input_request, metadata)
+                async for result_dict in result_generator:
+                    partial_response: str = await self._extract_tool_response_part(result_dict)
+                    if partial_response is not None:
+                        response_text = response_text + partial_response
+        except (asyncio.CancelledError, tornado.iostream.StreamClosedError):
+            # ensure generator is closed promptly
+            if result_generator is not None:
+                # Suppress possible exceptions: they are of no interest here.
+                with contextlib.suppress(Exception):
+                    await result_generator.aclose()
+                    result_generator = None
```

Comment:
Always, always close the underlying generator - that will ensure resources cleanup.